File: app/src/main/java/awersching/gammelgps/ui/MainActivity.java
package awersching.gammelgps.ui;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;

import awersching.gammelgps.R;
import awersching.gammelgps.location.Data;
import awersching.gammelgps.location.GPS;

import static awersching.gammelgps.util.Util.round;

public class MainActivity extends AppCompatActivity {

    private TextView currentSpeedTV;
    private TextView averageSpeedTV;
    private TextView maxSpeedTV;
    private TextView distanceTV;
    private TextView timeTV;

    private GPS gps;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        currentSpeedTV = findViewById(R.id.current_speed_tv);
        averageSpeedTV = findViewById(R.id.average_speed_tv);
        maxSpeedTV = findViewById(R.id.max_speed_tv);
        distanceTV = findViewById(R.id.distance_tv);
        timeTV = findViewById(R.id.time_tv);

        getPermissions();
        gps = new GPS(this);
    }

    @Override
    protected void onPause() {
        gps.pause();
        super.onPause();
    }

    @SuppressLint("CheckResult")
    @Override
    protected void onResume() {
        super.onResume();
        gps.start().subscribe(this::updateUI);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.exit_menu_item:
                gps.stop(false);
                finish();
                return true;
            case R.id.save_exit_menu_item:
                gps.stop(true);
                finish();
                return true;
        }
        return false;
    }

    private void getPermissions() {
        if (!permissionsGranted()) {
            String[] permissions = {
                    Manifest.permission.ACCESS_FINE_LOCATION,
                    Manifest.permission.WRITE_EXTERNAL_STORAGE
            };
            ActivityCompat.requestPermissions(this, permissions, 0);
        }
    }

    private boolean permissionsGranted() {
        boolean location = ActivityCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION) ==
                PackageManager.PERMISSION_GRANTED;
        boolean writeStorage = ActivityCompat.checkSelfPermission(this,
                Manifest.permission.WRITE_EXTERNAL_STORAGE) ==
                PackageManager.PERMISSION_GRANTED;
        return location && writeStorage;
    }

    private void updateUI(Data data) {
        currentSpeedTV.setText(round(data.getCurrentSpeed()));
        averageSpeedTV.setText(round(data.getAverageSpeed()));
        maxSpeedTV.setText(round(data.getMaxSpeed()));
        distanceTV.setText(round(data.getDistance()));
        timeTV.setText(data.getTime());
    }
}

File: app/src/main/java/awersching/gammelgps/ui/MainActivity.kt
package awersching.gammelgps.ui

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import awersching.gammelgps.R
import awersching.gammelgps.location.Data
import awersching.gammelgps.location.GpsService
import awersching.gammelgps.location.Round.round
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    private lateinit var gpsService: GpsService
    private val connection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            val binder = service as GpsService.GpsBinder
            gpsService = binder.get()
            val intent = Intent(this@MainActivity, GpsService::class.java)
            startForegroundService(intent)
            gpsService.locationUpdates().subscribe { data -> runOnUiThread { updateUI(data) } }
        }

        override fun onServiceDisconnected(name: ComponentName?) {
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        Permissions(this).request()
    }

    override fun onStart() {
        super.onStart()
        Intent(this, GpsService::class.java).also { intent ->
            bindService(intent, connection, Context.BIND_AUTO_CREATE)
        }
    }

    override fun onStop() {
        unbindService(connection)
        super.onStop()
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.exit_menu_item -> {
                gpsService.stop(false)
                finish()
                return true
            }
            R.id.save_exit_menu_item -> {
                gpsService.stop(true)
                finish()
                return true
            }
        }
        return false
    }

    private fun updateUI(data: Data) {
        currentSpeedTV.text = round(data.currentSpeed)
        averageSpeedTV.text = round(data.averageSpeed)
        maxSpeedTV.text = round(data.maxSpeed)
        distanceTV.text = round(data.distance)
        timeTV.text = data.time
    }
}
